Sources/SourceControl/GitRepository.swift

```diff
@@ -922,3 +963,153 @@ public struct GitCloneError: Error, CustomStringConvertible, DiagnosticLocationP
+            guard let progress = Double(matches[11]),
+                  let currentObjects = Int(matches[12]),
+                  let totalObjects = Int(matches[13]) else { return nil }
+
+            self = .resolvingDeltas(progress: progress / 100, currentObjects: currentObjects, totalObjects: totalObjects)
+
+        } else if matches[14] == "Receiving objects" {
+            guard let progress = Double(matches[15]),
+                  let currentObjects = Int(matches[16]),
+                  let totalObjects = Int(matches[17]) else { return nil }
+
+            let downloadProgress = matches[18]
+            let downloadSpeed = matches[19]
+
+            self = .receivingObjects(progress: progress / 100, currentObjects: currentObjects, totalObjects: totalObjects, downloadProgress: downloadProgress, downloadSpeed: downloadSpeed)
+
+        } else {
+            return nil
+        }
+    }
+
+    public var message: String {
+        switch self {
+        case .enumeratingObjects: return "Enumerating objects"
+        case .countingObjects: return "Counting objects"
+        case .compressingObjects: return "Compressing objects"
+        case .receivingObjects: return "Receiving objects"
+        case .resolvingDeltas: return "Resolving deltas"
+        }
+    }
+
+    public var step: Int {
+        switch self {
+        case .enumeratingObjects(let currentObjects):
+            return  currentObjects
+        case .countingObjects(_, let currentObjects, _):
+            return currentObjects
+        case .compressingObjects(_, let currentObjects, _):
+            return currentObjects
+        case .receivingObjects(_, let currentObjects, _, _, _):
+            return currentObjects
+        case .resolvingDeltas(_, let currentObjects, _):
+            return currentObjects
+        }
+    }
+
+    public var totalSteps: Int? {
+        switch self {
+        case .enumeratingObjects:
+            return 0
+        case .countingObjects(_, _, let totalObjects):
+            return totalObjects
+        case .compressingObjects(_, _, let totalObjects):
+            return totalObjects
+        case .receivingObjects(_, _, let totalObjects, _, _):
+            return totalObjects
+        case .resolvingDeltas(_, _, let totalObjects):
+            return totalObjects
+        }
+    }
+
+    public var downloadProgress: String? {
+        switch self {
+        case .receivingObjects(_, _, _, let downloadProgress, _):
+            return downloadProgress
+        default:
+            return nil
+        }
+    }
+
+    public var downloadSpeed: String? {
+        switch self {
+        case .receivingObjects(_, _, _, _, let downloadSpeed):
+            return downloadSpeed
+        default:
+            return nil
+        }
+    }
+
+    /// Processes stdout output and calls the progress callback with `GitStatus` objects.
+    static func gitFetchStatusFilter(_ bytes: [UInt8], progress: FetchProgress.Handler) {
```

Comment:
this feels out of place here, maybe lift back into GitRepository as a private function?

Reply:
`gitFetchStatusFilter` is called from both `GitRepositoryProvider` and  `GitRepository` so we could put it at the end of GitRepository.swift` as a `fileprivate func`. Maybe like this be29d1.